File: lifecycle_anslysis/system.py
import numpy as np

from lifecycle_anslysis.constants import OPEX_PER_YEAR, DRAM_WATTS_PER_256GB, GCI_CONSTANTS


class System:

    def __init__(self, die_size: float, performance_indicator: float, lifetime: int, dram_capacity: int,
                 ssd_capacity: int, hdd_capacity: int, cpu_tdp: int) -> None:
        """
        :param die_size: in cm^2
        :param performance_indicator:
        :param lifetime: in years
        :param dram_capacity: in GB
        :param ssd_capacity: in GB
        :param hdd_capacity: in GB
        :param cpu_tdp: in Watt
        """
        self.packaging_size = die_size
        self.specint = performance_indicator
        self.lifetime = lifetime
        self.dram_capacity = dram_capacity
        self.ssd_capacity = ssd_capacity
        self.hdd_capacity = hdd_capacity
        self.cpu_tdp = cpu_tdp

    def calculate_capex_emissions(self):
        ####### Source of the constants: https://ugupta.com/files/Gupta_ISCA2022_ACT.pdf

        MPA = 0.5  ### Procure materials | kg co2 per cm2
        EPA = 2.15  ### 0.8-3.5 | Fab Energy | kWh per cm2
        CI_fab = 0.365  ### 30-700  | g co2 per kWh  --> converted to kg co2 per kWh
        GPA = 0.3  ### 0.1-0.5 | Kg CO2 per cm2
        _yield = 0.875  ### 0-1  | Fab yield
        E_DRAM = 0.3  ### 0 - 0.6 | Kg CO2/GB
        E_SDD = 0.015  ### 0 - 0.03 | Kg CO2/GB
        E_HDD = 0.06  ### 0 - 0.12 | Kg CO2/GB

        # Note assume package size == die size
        capex_cpu = ((CI_fab * EPA + GPA + MPA) * self.packaging_size) / _yield  #### Kg Co2
        capex_dram = self.dram_capacity * E_DRAM  #### Kg Co2
        capex_ssd = self.ssd_capacity * E_SDD  #### Kg Co2
        capex_hdd = self.hdd_capacity * E_HDD  #### Kg Co2

        capex_total = capex_cpu + capex_dram + capex_ssd + capex_hdd  #### Kg Co2

        return capex_total

    def generate_accumm_projected_opex_emissions(self, time_horizon: int, system_id: str, country: str,
                                                 utilization: float, lookup: bool):
        if lookup:
            opex_per_year = OPEX_PER_YEAR[country][utilization][system_id]
        else:
            opex_per_year = self.calculate_opex_emissions(utilization, country)
        projected_emissions = [i * opex_per_year for i in range(1, time_horizon + 1)]

        return np.array(projected_emissions)

    def calculate_opex_emissions(self, utilization: float, country: str):
        ######## Source of GCI: https://app.electricitymaps.com/zone/DE --> 2023 average for DE

        cpu_energy_consumption = (self.cpu_tdp * (utilization / 100)) / 1000  #### kW
        dram_energy_consumption = ((self.dram_capacity / 256) * DRAM_WATTS_PER_256GB) / 1000  #### kW

        # Watts according to https://www.ssstc.com/knowledge-detail/ssd-vs-hdd-power-efficiency/#:~:text=On%20average%2C%20SSDs%20consume%20around,may%20consume%203%2D4%20watts.
        ssd_energy_consumption = (3 if (self.ssd_capacity > 0) else 0) / 1000  ###kW
        hdd_energy_consumption = (7 if (self.hdd_capacity > 0) else 0) / 1000  ###kW

        total_watts = cpu_energy_consumption + dram_energy_consumption + ssd_energy_consumption + hdd_energy_consumption
        total_watts_per_year = 24 * 7 * 52 * total_watts  ### kWh
        GCI = GCI_CONSTANTS[country]

        OPEX = total_watts_per_year * GCI  ###### Kg co2 per year

        return OPEX

File: frontend/src/utility/lifecycle_analysis/system.ts
import * as constants from './constants.ts';
import { GRID_INTENSITY } from '../../assets/grid_intensities.ts';
import { Country } from '../../assets/grid_intensities.ts';

export const CPU = 'CPU';
export const RAM = 'RAM';
export const SSD = 'SSD';
export const HDD = 'HDD';
export type Components = typeof CPU | typeof RAM | typeof SSD | typeof HDD;

export interface CapexType extends Record<Components, number> {
  TOTAL: number;
}

export interface OpexType extends Record<Components, number> {
  TOTAL: number;
  opexPerYear: number;
}

export interface ProjectedOpexEmissionsType {
  projected: number[];
  opexBreakdown: OpexType;
}

export class System {
  packagingSize: number;
  performanceIndicator: number;
  lifetime: number;
  dramCapacity: number;
  ssdCapacity: number;
  hddCapacity: number;
  cpuTdp: number;

  constructor(
    dieSize: number,
    performanceIndicator: number,
    lifetime: number,
    dramCapacity: number,
    ssdCapacity: number,
    hddCapacity: number,
    cpuTdp: number
  ) {
    /**
     * @param dieSize in cm^2
     * @param performanceIndicator
     * @param lifetime in years
     * @param dramCapacity in GB
     * @param ssdCapacity in GB
     * @param hddCapacity in GB
     * @param cpuTdp in Watts
     */
    this.packagingSize = dieSize;
    this.performanceIndicator = performanceIndicator;
    this.lifetime = lifetime;
    this.dramCapacity = dramCapacity;
    this.ssdCapacity = ssdCapacity;
    this.hddCapacity = hddCapacity;
    this.cpuTdp = cpuTdp;
  }

  calculateCapexEmissions(): CapexType {
    // Constants
    const MPA = 0.5; // Procure materials | kg CO2 per cm^2
    const EPA = 2.15; // Fab Energy | kWh per cm^2
    const CI_FAB = 0.365; // g CO2 per kWh converted to kg CO2 per kWh
    const GPA = 0.3; // Kg CO2 per cm^2
    const FAB_YIELD = 0.875; // Fab yield
    const E_DRAM = 0.3; // Kg CO2/GB
    const E_SSD = 0.015; // Kg CO2/GB
    const E_HDD = 0.06; // Kg CO2/GB

    // Calculate emissions
    const capexCpu =
      ((CI_FAB * EPA + GPA + MPA) * this.packagingSize) / FAB_YIELD;
    const capexDram = this.dramCapacity * E_DRAM;
    const capexSsd = this.ssdCapacity * E_SSD;
    const capexHdd = this.hddCapacity * E_HDD;

    return {
      CPU: capexCpu,
      RAM: capexDram,
      SSD: capexSsd,
      HDD: capexHdd,
      TOTAL: capexCpu + capexDram + capexSsd + capexHdd
    }
  }

  generateAccumProjectedOpexEmissions(
    timeHorizon: number,
    systemId: string,
    country: Country,
    utilization: number,
    opexCalculation: string
  ): ProjectedOpexEmissionsType {
    let opexPerYear: number;

    let opexBreakdown: OpexType;

    // never reaching this if statement, want to delete but too much refactoring
    // TODO: later
    if (opexCalculation === constants.HPE_POWER_ADVISOR) {
      opexPerYear = constants.OPEX_PER_YEAR[country][utilization][systemId];
    };

    opexBreakdown = this.calculateOpexEmissions(utilization, country);
    opexPerYear = opexBreakdown.opexPerYear;

    const projected =  Array.from({ length: timeHorizon }, (_, i) =>
      i * opexPerYear
    );

    return {
      projected,
      opexBreakdown
    }
  }

  generateNormalizedPowerUsage(utilization: number): number {
    // Constants
    const powerSlope = (100 - 50) / (100 - 0);
    const intercept = 50;

    return (intercept + utilization * powerSlope) / 100;
  }

  calculateOpexEmissions(utilization: number, country: Country): OpexType {
    const normalizedPowerUsage = this.generateNormalizedPowerUsage(utilization);

    // Energy consumption calculations
    const cpuEnergyConsumption = (this.cpuTdp * normalizedPowerUsage) / 1000; // kW
    const dramEnergyConsumption =
      ((this.dramCapacity / 256) * constants.DRAM_WATTS_PER_256GB) / 1000; // kW
    const ssdEnergyConsumption =
      (this.ssdCapacity > 0 ? 3 : 0) / 1000; // kW
    const hddEnergyConsumption =
      (this.hddCapacity > 0 ? 7 : 0) / 1000; // kW

    const totalWatts =
      cpuEnergyConsumption +
      dramEnergyConsumption +
      ssdEnergyConsumption +
      hddEnergyConsumption;
    const totalWattsPerYear = 24 * 7 * 52 * totalWatts; // kWh
    const GCI = (GRID_INTENSITY[country] || 0) / 1000;

    return {
      CPU: cpuEnergyConsumption,
      RAM: dramEnergyConsumption,
      SSD: ssdEnergyConsumption,
      HDD: hddEnergyConsumption,
      TOTAL: totalWatts,
      opexPerYear: totalWattsPerYear * GCI
    }
}
}
